niimpy/EDA/EDA_punchcard.py

```diff
@@ -7,6 +7,117 @@
 import pandas as pd
 import plotly.express as px
 
+def get_timerange_(df,resample):
+    """get first and last timepoint from the dataframe, 
+    and return a resampled datetimeindex.
+    
+    Parameters
+    ----------
+    df : Pandas Dataframe
+        Dataframe containing the data
+    ressample : str
+        Resample parameter e.g., 'D' for resampling by day
+
+    Returns
+    -------
+    date_index : pd.DatatimeIndex
+        Resampled DatetimeIndex
+    """
+    resample_options = ['D','H']
+    
+    assert isinstance(df,pd.DataFrame), "df is not a pandas dataframe"
+    assert isinstance(resample,str), "resample is not a string"
+    assert (resample in resample_options), f"resample option: {resample} is not valid. Available options: {resample_options}."
+    
+    start = df.index.min()
+    end = df.index.max()
+    
+    if resample == 'D':
+        date_index = pd.date_range(start = start.strftime('%Y-%m-%d'), end = end.strftime('%Y-%m-%d'),freq='D')
+    if resample == 'H':
+        date_index = pd.date_range(start = start.strftime('%Y-%m-%d-%H'), end = end.strftime('%Y-%m-%d-%H'),freq='H')
+            
+    return date_index
+
+def combine_dataframe_(df,user_list,columns,res,date_index):
+    """resample values from multiple users into new dataframe
+    
+    Parameters
+    ----------
+    df : Pandas Dataframe
+        Dataframe containing the data
+    user_list : list
+        List containing user names/id's (str)
+    columns : list
+        List of column names (str) to be plotted
+    res : str
+        Resample parameter e.g., 'D' for resampling by day
+    date_index : pd.date_range
+        Date range used as an index
+    
+    Returns
+    -------
+    df_comb : pd.DataFrame
+        Resampled and combined dataframe
+    """
+    assert isinstance(df,pd.DataFrame), "df is not a pandas dataframe."
+    assert isinstance(user_list,list), "user_list is not a list."
+    assert isinstance(columns, list), "columns is not a list"
+    assert isinstance(res,str), "res is not a string."
+    assert isinstance(date_index,pd.core.indexes.datetimes.DatetimeIndex), "date_index is not a DatetimeIndex."
+    
+    df_comb = pd.DataFrame(index=date_index)
+    df_comb.index = pd.to_datetime(df_comb.index)
+
+    for u in user_list:
+        df_temp = df[df['user'] == u][columns].resample(res).mean()
```

Comment:
can we generalize the sampling method? In cor:ona dataset for example, sometimes the cumulative data is required, so the sampling method could also be `sum`.

Reply:
Generalized the function, uses user-defined aggregation function.